Modify prompt using saist.rules

X3NOSAURUS opened the pull request:
- Created PromptRules class to read saist.rules from the working directory
- PROMPT_OVERRIDE to fully replace the prompt
- PROMPT_PRE and PROMPT_POST to modify prompt before/after
- Integrated apply_rules into analyze_single_file to adjust the prompt
- Falls back to original prompt if saist.rules doesn't exist or is empty

Commit:
- Created PromptRules class to read saist.rules from the working directory
- PROMPT_OVERRIDE to fully replace the prompt
- PROMPT_PRE and PROMPT_POST to modify prompt before/after
- Integrated apply_rules into analyze_single_file to adjust the prompt
- Falls back to original prompt if saist.rules doesn't exist or is empty

SimonGurney commented on saist/main.py:
```diff
@@ -42,9 +44,9 @@ async def analyze_single_file(scm: Scm, adapter: BaseLlmAdapter, filename, patch
     """
     system_prompt = prompts.DETECT
```
It's the system_prompt which needs the tailored rules applying, not the user prompt which contains the file contents

SimonGurney commented on saist/util/rules.py:
```diff
@@ -0,0 +1,39 @@
+
+        override_prompt = rules.get("PROMPT_OVERRIDE")
+        if override_prompt:
+            #skip pre and post if override is present
```
I don't think this should be the desired behaviour. If the user wants to declare a PRE and POST with their OVERRIDE, we should probably use it

SimonGurney commented on saist/util/rules.py:
```diff
@@ -0,0 +1,39 @@
+    @staticmethod
+    def load_rules():
+        if not os.path.exists(PromptRules.RulesFile):
+            print("No saist.rules file found.")
```
Take a look at how we do logging in the codebase and amend these print statements. Our logging is pretty poorly standardised, but yo should find a correct way of doing it

SimonGurney commented on saist/util/rules.py:
I wonder if we should log which of the 3 keys (or multiple) we detected?

SimonGurney commented:
Consider updating the docs, and adding a saist.rules example file?

Commit:
- Applies rules to system_prompt instead of user prompt
- PRE and POST no longer skipped when using OVERRIDE
- Standardised logging
- Added example rules file
- Updated README

Commit:
- Changed logger name to use a module-level logger

X3NOSAURUS commented on saist/util/rules.py:
```diff
@@ -0,0 +1,43 @@
+import yaml
+import logging
+
+logger = logging.getLogger("saist")
```
thanks for this! just had a better look at logging and naming conventions, makes sense now

SimonGurney commented:
Check how the README renders...

Commit:
-Closed yaml block on README

SimonGurney commented:
This is looking better, but I've had a quick review and the DETECT prompt is already split into PRE and POST in the prompts class. This should probably be tidied up...

Commit:
- Amended apply_rules to work properly with prompts
- Changed system prompt definition within analyze_single_file to match the updated logic

Commit:
-Ammended apply rules to work as intended
-Ammened system prompt back to DIRECT for simplicity
- Moved return out of else in load_rules

X3NOSAURUS commented:
> This is looking better, but I've had a quick review and the DETECT prompt is already split into PRE and POST in the prompts class. This should probably be tidied up...

Apologies for the delay in sorting this one properly - I realised my first go at it was a mess!

SimonGurney commented:
Hmmm, it now looks like you don't provide the detect post part of the prompt to the system_prompt at all?

X3NOSAURUS commented:
> Hmmm, it now looks like you don't provide the detect post part of the prompt to the system_prompt at all?

I was just using DETECT as this combines the pre and post within the prompts class - In my apply rules func I combine the pre and post (with the override if there is one). I can pass pre and post separately if that would be preferred?

Commit:
-Ammended apply rules again, my last itteration was overwriting the original prompt with the new pre and post. I don't think this is the desired intention.
-final_prompt should now contain pre and post (if they are present) wrapping either the original detect prompt or the overridden detect prompt
